### DISCUSSION TITLE
When using WithPagination trait, it removes trailing slash from URLs

### DISCUSSION THREAD

**Discussion by adfriend2003** [livewire/livewire#4503] [1. Help] [answered]
I am trying to persist trailing slashes in my URLs so they looks like the following:
`http://www.example.com/`
`http://www.example.com/products/`
`http://www.example.com/products/?page=1`

However, when I am using `WithPagination` trait and defines `protected $queryString = ['page'];` then as soon my page loads the trailing slash gets removed.

**To produce this behavior:**
In component class file add following two lines and try to visit it with URL which has trailing slash.
1) `Use WithPagination;`
2) `protected $queryString = ['page'];`

I have commented out following lines from `.htaccess` file so I can use trailing slashes in my URLs:
```
# Redirect Trailing Slashes If Not A Folder...
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_URI} (.+)/$
RewriteRule ^ %1 [L,R=301]
```

Is there anyway to persist trailing slashes like as I mentioned above?

Thank you.

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

**Comment by joshhanley** [comment]
@adfriend2003 I think the query string system probably removes the trailing slash (although I'm not 100% sure on that). How come you are trying to enforce the trailing slash?

**Reply by adfriend2003** [reply]
I only commented out few lines of `.htaccess` file (as I mentioned in my question) so that Laravel don't remove trailing slashes from the URLs.

The trailing slashes stays if I don't use `WithPagination` trait or `protected $queryString = ['page'];`

Edit: Another thing if I use `WithPagination` trait and add following function in the component class then **query string parameters**  get removed from the URL (if there is any) but trailing slashes stays.

```
class Products extends Component 
{
  ...
  public function getQueryString() {
      return [];
  }
  ...
}
```

**Comment by adfriend2003** [comment]
On further digging it looks like, when Livewire is going to set the `$queryString` variables using browser APIs' either `window.history.replaceState` or `window.history.updateState` then by that time it omits that trailing slashes from the URLs.

**Comment by joshhanley** [comment]
@adfriend2003 yeah it does. Can you explain why you are trying to enforce the trailing slash?

**Reply by adfriend2003** [reply]
We have an old site which is using trailing slashes at the end of some URLs. So, we are trying the same here as we don't want to de-rank our URLs on Google.

**Reply by joshhanley** [reply]
@adfriend2003 ah right, wow didn't know a trailing slash would impact that! Yeah I'm not sure how you can work around it without PRing a change to Livewire.

**Comment by flauntsam** [comment, marked as answer]
Hey, we just had this problem on a client's website that we are working on (Similar situtation of the old website used trailing slashes for urls). 

We managed to fix the problem by adjusting the livewire.js. These was the steps that I took (Just in case it is helpful for anyone)

I ran `php artisan livewire:publish --assets` to bring the js files into the repo.

Then opened public/vendor/livewire/livewire.js and changed `onlyChangeThePathAndQueryString` from 
`
function onlyChangeThePathAndQueryString(url) {
           if (url) {
               var destination = new URL(url),
                   afterOrigin = destination.href.replace(destination.origin, "").replace(/\?$/, "");
               
               return window.location.origin + afterOrigin + window.location.hash;
           }
       }
`
to

`
function onlyChangeThePathAndQueryString(url) {
           if (url) {
               var destination = new URL(url),
                   afterOrigin = destination.href.replace(destination.origin, "").replace(/\?$/, "");
               
               afterOrigin = window.location.pathname.substring(1);
               
               return window.location.origin + afterOrigin + window.location.hash;
           }
       }
`

This change was adding in `afterOrigin = window.location.pathname.substring(1);` and this seemed to have fixed the problem.

**Reply by adfriend2003** [reply]
It worked for me, thank you.
